cirq/work/observable_measurement_data.py

```diff
@@ -0,0 +1,410 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import dataclasses
+import datetime
+import itertools
+from typing import Dict, List, Tuple, \
+    TYPE_CHECKING
+
+import numpy as np
+
+from cirq import protocols
+from cirq.work.observable_settings import InitObsSetting, _max_weight_observable, _max_weight_state, \
+    _MeasurementSpec
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def _get_real_coef(observable: 'cirq.PauliString'):
+    """Assert that a PauliString has a real coefficient and return it."""
+    coef = observable.coefficient
+    if not np.isclose(coef.imag, 0):
+        raise ValueError(f"{observable} has a complex coefficient.")
+    return coef.real
+
+
+def _obs_vals_from_measurements(bitstrings: np.ndarray,
+                                qubit_to_index: Dict['cirq.Qid', int],
+                                observable: 'cirq.PauliString'):
+    """Multiply together bitstrings to get observed values of operators."""
+    idxs = [qubit_to_index[q] for q in observable.keys()]
+    # Make sure any unit8's coming in get a sign. eigenvalues will be +- 1.
+    # Select the columns we need for this observable
+    selected_bitstrings = np.asarray(bitstrings[:, idxs], dtype=np.int8)
+    # Transform bits to eigenvalues; ie (+1, -1)
+    selected_obsstrings = 1 - 2 * selected_bitstrings
+    coeff = _get_real_coef(observable)
+    obs_vals = coeff * np.prod(selected_obsstrings, axis=1)
+    return obs_vals
+
+
+def _stats_from_measurements(bitstrings: np.ndarray,
+                             qubit_to_index: Dict['cirq.Qid', int],
+                             observable: 'cirq.PauliString',
+                             ) -> Tuple[float, float]:
+    """Return the mean and squared standard error of the mean for the given
+    observable according to the measurements in `bitstrings`."""
+    obs_vals = _obs_vals_from_measurements(
+        bitstrings, qubit_to_index, observable)
+    obs_mean = np.mean(obs_vals)
+
+    # Terminology: This isn't technically the variance. It's the
+    # standard error of the mean, but squared.
+    obs_err = np.var(obs_vals) / len(obs_vals)
+    return obs_mean.item(), obs_err.item()
+
+
+@protocols.json_serializable_dataclass(frozen=True)
+class ObservableMeasuredResult:
+    """The result of an observable measurement.
+
+    Please see `flatten_grouped_results` or
+    `BitstringAccumulator.results` for information on how to get these
+    from `measure_observables` return values.
+
+    This is a flattened form of the contents of a `BitstringAccumulator`
+    which may group many simultaneously-observable settings into one object.
+    As such, `BitstringAccumulator` has more advanced support for covariances
+    between simultaneously-measured observables which is dropped when you
+    flatten into these objects.
+    """
```

Comment:
The fields need docstrings as part of the class docstring.

Reply:
done